### PR TITLE
Add support for multiple tensor input/output

### PR DESCRIPTION
**Pull request by FynnBe**
This PR adds support for models with multi tensor input and output as the bioimageio RDF defines it.

side note for further/later discussion/development:
`PredictionPipeline` somewhat unnecessarily redefines some parts of  the bioimageio.core.nodes definitions with named/tagged shapes, etc. 
I don't think this slight difference is really justifying the whole redifinition of many fields; instead the implementation of the nodes could be changed or the axis naming, tagging left out here (the axes names are always accessible!). As this would be a bigger change, it is left untouched here and locally `NamedImplicitOutputShape` is defined as a tagged equivalent to `nodes.ImplicitOutputShape`.

### CODE REVIEW

**Comment by constantinpape** [comment]
> I don't think this slight difference is really justifying the whole redifinition of many fields; instead the implementation of the nodes could be changed or the axis naming, tagging left out here (the axes names are always accessible!). As this would be a bigger change, it is left untouched here and locally `NamedImplicitOutputShape` is defined as a tagged equivalent to `nodes.ImplicitOutputShape`.

Indeed we should get rid of the duplication, but I agree that it's better to do this in a follow up PR.

**Review by constantinpape — COMMENTED** [review]
Looks good overall. Can you add a test for pytorch using the new example model?
(If you want I can also add the multi tensor version of onnx/torchscript, so that we can add tests for these.)

**Comment by FynnBe** [comment]
Alright working on the multi tensor modifications...

Constantin Pape ***@***.***> schrieb am So., 12. Sep. 2021,
20:01:

File: bioimageio/core/prediction_pipeline/_model_adapters/_keras_model_adapter.py [hunk 1]
```diff
@@ -10,19 +10,18 @@
...
-        self.devices = []
 
-        weight_file = self.spec.weights["keras_hdf5"].source
-        self.model = keras.models.load_model(weight_file)
+        weight_file = bioimageio_model.weights["keras_hdf5"].source
+        self._model = keras.models.load_model(weight_file)
+        self._output_axes = [tuple(out.axes) for out in bioimageio_model.outputs]
+
+    def forward(self, *input_tensors: xr.DataArray) -> List[xr.DataArray]:
+        result = self._model.predict(*[ipt.data for ipt in input_tensors])
```
[the remark this answers is not in the document]
Comment (FynnBe) [reply bioimageio/core/prediction_pipeline/_model_adapters/_keras_model_adapter.py:22]: it seems like the keras model does not accept xr.DataArrays, therefore this list comprehensions extracts the numpy array.
```RuntimeError: Input must be a list of dictionaries or a single numpy array for input 'input.1'.```

[the remark this answers is not in the document]
Comment (FynnBe) [reply bioimageio/core/prediction_pipeline/_model_adapters/_keras_model_adapter.py:22]: no, I think I've mistaken the error message from another test failure...

**Comment by FynnBe** [comment]
tests using the new example models are still missing...

File: bioimageio/core/prediction_pipeline/_model_adapters/_keras_model_adapter.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (constantinpape) [reply bioimageio/core/prediction_pipeline/_model_adapters/_keras_model_adapter.py]: > it seems like the keras model does not accept xr.DataArrays, therefore this list comprehensions extracts the numpy array.

Ah ok, sorry I missed the `ipt.data`.

File: bioimageio/core/prediction.py [hunk 2]
```diff
@@ -191,11 +189,31 @@ def get_tiling(shape, tile_shape, halo, input_axes):
...
+    inputs: List[xr.DataArray],
+    outputs: List[xr.DataArray],
+    tile_shapes: List[dict],
+    halos: List[dict],
+):
+    if len(inputs) > 1:
+        raise NotImplementedError("Tiling with multiple inputs not implemented yet")
+
+    if len(outputs) > 1:
+        raise NotImplementedError("Tiling with multiple inputs not implemented yet")
```
Comment (constantinpape) [inline bioimageio/core/prediction.py:203]:
```suggestion
        raise NotImplementedError("Tiling with multiple outputs not implemented yet")
```

File: bioimageio/core/prediction.py [hunk 3]
```diff
@@ -211,75 +229,85 @@ def load_tile(tile):
...
-    axes = tuple(prediction_pipeline.input_axes)
-    tagged_data = [xr.DataArray(ipt, dims=axes) for ipt in inputs]
+
+    tagged_data = [xr.DataArray(ipt, dims=axes) for ipt, axes in zip(inputs, prediction_pipeline.input_axes)]
     return prediction_pipeline.forward(*tagged_data)
 
 
-def predict_with_padding(prediction_pipeline, inputs, padding, pad_right=True):
-    if isinstance(inputs, (np.ndarray, xr.DataArray)):
+def predict_with_padding(prediction_pipeline, inputs, padding, pad_right=True) -> List[xr.DataArray]:
```
Comment (constantinpape) [inline bioimageio/core/prediction.py:250]: I think we should also enable passing padding as a list here; if different input tensors are given they might also need to be padded differently. 
So I would add something like `if not isinstance(padding, (tuple, list)): padding = [padding]` and then also iterate over padding in line 255.

File: bioimageio/core/prediction.py [hunk 4]
```diff
@@ -211,75 +229,85 @@ def load_tile(tile):
...
+    )
 
-    predict_with_tiling_impl(prediction_pipeline, input_, output, tile_shape, halo, input_axes)
-    return output
+    return outputs
 
 
 def parse_padding(padding, model):
+    if len(model.inputs) > 1:
+        raise NotImplementedError("Tiling for multiple inputs not yet implemented")
```
Comment (constantinpape) [inline bioimageio/core/prediction.py:310]:
```suggestion
        raise NotImplementedError("Padding for multiple inputs not yet implemented")
```

**Review by constantinpape — COMMENTED** [review]
I left a few comments and suggestions, looks good otherwise. But let's implement prediction tests using the multi-tensor model before merging this.

**Comment by FynnBe** [comment]
`test_prediction_pipeline_torch` for the new multi tensor model fails with unequal outputs. the first is:
output: 
image

expected:
image

**Comment by constantinpape** [comment]
@FynnBe indeed I forgot to update the test outputs after making some changes to the models and weights to fix another issue.
I updated the outputs now. Can you give it another try?

**Pull request merged**